File: p337.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    unordered_map<TreeNode*, int> solution;
public:
    int rob(TreeNode* root) {
        if(root == NULL)
            return 0;
        else if(solution[root])
            return solution[root];
        int option1 = rob(root->left) + rob(root->right);
        int option2 = root->val;
        if(root->left)
            option2 += rob(root->left->left) + rob(root->left->right);
        if(root->right)
            option2 += rob(root->right->left) + rob(root->right->right);
        solution[root] = max(option1, option2);
        return solution[root];
    }
};

File: p337.py
from functools import lru_cache
from DS.TreeNode import TreeNode

class Solution:
    def rob(self, root: TreeNode) -> int:
        
        @lru_cache(None)
        def dfs(node):
            if node is None:
                return 0
            
            ans = node.val
            if node.left:
                ans += dfs(node.left.left) + dfs(node.left.right)
            if node.right:
                ans += dfs(node.right.left) + dfs(node.right.right)
                
            return max(ans, dfs(node.left) + dfs(node.right))
        
        return dfs(root)